feat(docs-infra): Add opensearch description

s-gbz opened the pull request:
Enables Chrome users to search angular.io from the browsers navigation bar.
Not sure if compatible with Firefox yet.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features) <- **Not sure how or if necessary, since it's a doc feature**
- [X] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[X] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #24107


## What is the new behavior?


## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
@IgorMinar s checklist:

> ensure that we can activate the search via a url by navigating to angular.io/?q=

[X] done by angular.io/?search=<query>
> we might consider different url scheme for example that used by auxiliary routes in the Angular router > or create a dedicated /search page - this needs to be thought through

[X] I went for a non-dedicated page
>if we go with non-dedicated page and we just overlay the search results over e.g. the homepage, we need to ensure that we clean up the url once the search result overlay is closed

[X] URL is reduced to angular.io after search result overlay is closed

> we also need to think a bit about how to report these hits in analytics, also see #23581

[ ] Not sure how Angulars analytics work

mary-poppins commented:
You can preview cb2510c at https://pr25479-cb2510c.ngbuilds.io/.

IgorMinar commented on aio/src/assets/opensearch.xml:
```diff
@@ -0,0 +1,9 @@
+<InputEncoding>UTF-8</InputEncoding>
+<Image width="16" height="16" type="image/x-icon">../assets/images/favicons/favicon-16x16.png</Image>
+<Url type="text/html" method="get" template="https://angular.io/?search={searchTerms}"/>
+<moz:SearchForm>https://angular.io/?search={searchTerms}</moz:SearchForm>
```
can these angular.io urls be relative so that the search works on all the subdomains?

for example if I type next.angular.io into the address bar, hit tab, type a query, I should see the results on next.angular.io.

s-gbz replied:
Hello Igor,
opensearch handles URL paremeters very picky. Testing it localy only succeeds with workarounds, since opensearch doesn't accept port specification like "localhost:4200" (This made the PR way harder than it should've been)

So no, you can't set the URL to a relative directory like a dot in a unix file system, but subdomains *should* work if they share the same webstorage as the main domain. 

[This](https://stackoverflow.com/questions/18642586/how-to-define-a-trigger-keyword-in-opensearchdescription-xml-file) and [this](https://meta.stackexchange.com/questions/252701/dont-include-an-opensearch-provider-on-nonexistent-subdomains) question suggest that it *should* work though.

s-gbz replied:
The various files have now been added and moddified except for the basic opensearch.xml file which still contains angular.io as URL, because as previously mentioned, opensearch does not support port specifications in an URL.

If you really wanted the file to include some local adress, the only possibility here would be to set the URL to 127.0.0.1 (or localhost) and simply check if it's addable as a search engine. We still wouldn't be able to use it for searching though.

IgorMinar commented on aio/src/app/search/search-box/search-box.component.ts:
```diff
@@ -58,6 +59,13 @@ export class SearchBoxComponent implements OnInit {
+   */
+  focusOut() {
+    this.locationService.setSearch('', { undefined });
+  }
```
this doesn't appear to work for me and I think that it should be tied to search results closing and not the search box loosing focus.

s-gbz replied:
Correct me if I'm wrong, but you do close the search results by pressing escape or clicking somewhere outside which results in the search bar loosing focus.

You can reproduce the result like so:
`http://localhost:4200?search=pipe` -> the search "popup"? opens
Click outside -> URL is reduced to `http://localhost:4200`. 

If it doesn't you might have to click on the bar and click outside again.
I had tried something like onClose previously, but unfortunately without success.

What's your take on it? 
[Here's](https://christianliebel.com/2016/05/angular-2-a-simple-click-outside-directive/) a possible JS workaround.

IgorMinar replied:
There are two issues with relying on the search box focus:

- if you use keyboard
  - go to https://pr25479-cb2510c.ngbuilds.io/?search=directive
  - hit esc
  - you'll see that the result page refreshes with the search box having focus (and url not being updated)
  - to get to the right state you have to hit esc again
  - now the result page closes and url is updated

- if you use mouse/pointer/tap
  - go to https://pr25479-cb2510c.ngbuilds.io/?search=directive
  - click/tap anywhere outside of the search result
  - the result disappears but because the search box never got and lost focus the url is not updated

IgorMinar reviewed (changes requested):
thanks for the PR, you seem to be heading in the right direction but this PR needs more work.

gkalpak commented:
> By the way, would you prefer to have the commits squashed into one?

Nooooo... 😛
For reviewing, it is much better to not touch old commits and create new commits for any modifications. Once the PR is ready for review, you can squash the commits.
Even better, you can create `--fixup` commits, which our merge script knows to squash when merging, so you don't have to manually squash.

gkalpak commented on aio/src/app/app.component.ts:
```diff
@@ -384,6 +384,7 @@ export class AppComponent implements OnInit {
 
   hideSearchResults() {
     this.showSearchResults = false;
+    this.locationService.setSearch('', { undefined });
```
Why this change? It is unrelated to this PR (afaict) and not desirable (imo).
One thing that could be improved, is to select-all in the search box on focus.

s-gbz replied:
@IgorMinar requested to have the URL query cleaned when closing search results.

E.g. 
1. angular.io/?search=pipe 
2. Search results for pipe appear
3. User closes search results
4. URL is angular.io again

PS: I learned a handful of abbreviations by looking into your comments, thanks for that 😄

gkalpak replied:
Hm...I see. That's not how navigating to `/?search=...` works today (and deliberately so, I think).
(BTW, @s-gbz, the `{ undefined }` part seems wrong.)

I am fine removing the `search=...` part if @IgorMinar feels like it :smiley:

s-gbz replied:
> I am fine removing the search=... part if @IgorMinar feels like it 😃

What's the issue about using the `search` query? This PR is gladly making use of it just as [documented](https://github.com/angular/angular/blob/master/aio/src/app/search/search-box/search-box.component.ts#L9-L10) here.

> (BTW, @s-gbz, the { undefined } part seems wrong.)

The [method](https://github.com/angular/angular/blob/master/aio/src/app/shared/location.service.ts#L84) accepts undefined as a parameter.
Can you give me more insight why it might be wrong and in case it is how to fix it? 

Or maybe there is another way to reset the URL *without* reloading the page. Unfortunately this is how `location.replace()` behaves.

gkalpak replied:
>> I am fine removing the `search=...` part if @IgorMinar feels like it 😃

> What's the issue about using the `search` query? This PR is gladly making use of it just as [documented](https://github.com/angular/angular/blob/master/aio/src/app/search/search-box/search-box.component.ts#L9-L10) here.

No, I meant I am fine with removing `search=...` from the URL when closing search results.

##
>> (BTW, @s-gbz, the `{ undefined }` part seems wrong.)

> The method accepts undefined as a parameter.
> Can you give me more insight why it might be wrong and in case it is how to fix it?

Actually, the method does not accept `undefined` (it accepts an object whose values can be strings or undefined). And you are not passing `undefined`, you are passing `{ undefined }`, which is shorthand for `{undefined: undefined}` (i.e. an object with an `'undefined'` key, which happens to have `undefined` as value).

This will (accidentally) have the desired result (since keys with `undefined` values are ignored), but is confusing. You can just use `locationService.setSearch('', {})`.
(Alternatively, you could update the `setSearch()` method to accept `undefined` as the second argument.)

s-gbz replied:
Okay thank you, I will prepare a new commit now.

gkalpak commented on aio/src/assets/opensearch.archive.xml:
```diff
@@ -0,0 +1,9 @@
+<Description>Search Angular</Description>
+<InputEncoding>UTF-8</InputEncoding>
+<Image width="16" height="16" type="image/x-icon">../assets/images/favicons/favicon-16x16.png</Image>
+<Url type="text/html" method="get" template="https://v6.angular.io/?search={searchTerms}"/>
```
The archive URL should not be hard coded. We have several archived versions.
Maybe we could have a generic file template and replace the domain at [build time][1] (where we know the [domain][2]).

[1]: https://github.com/angular/angular/blob/3e661721968f841c8d835318c0cddf1fa1f29646/aio/scripts/deploy-to-firebase.sh#L86-L104
[2]: https://github.com/angular/angular/blob/3e661721968f841c8d835318c0cddf1fa1f29646/aio/scripts/deploy-to-firebase.sh#L71

s-gbz replied:
Igor previously asked me to prepare several files for the angular-cli file replacement (which I've done).

So do you mean a generic file template that 
- specificaly replaces the archived URL 
or
- substitutes the various opensearch files?

If we decide to that, is [this](https://www.packtpub.com/mapt/book/networking_and_servers/9781849513760/4/ch04lvl1sec62/replacing-strings-from-a-text-or-file) the right way?

gkalpak replied:
I mean a generic file template, which would contain a placeholder for the domain that we would replace with the actual domain during deployment (in [deploy-to-firebase.sh][1]).
We could do the same in [deploy-preview.sh][2] if we want this to be available for previews as well.

This could be a yarn script that gets passed the domain and copies the generic template into the dist directory after replacing the domain placeholder with the actual domain. (We could call this yarn script from wherever we need it (even as a `postbuild` script to have it during local development).

How this script is implemented (`sed`, JS, etc.) is irrelevant.

[1]: https://github.com/angular/angular/blob/3e661721968f841c8d835318c0cddf1fa1f29646/aio/scripts/deploy-to-firebase.sh#L86-L104
[2]: https://github.com/angular/angular/blob/3e661721968f841c8d835318c0cddf1fa1f29646/aio/scripts/deploy-preview.sh

s-gbz replied:
This is a good idea and nice to know, I will look into that. 

It's great how much there's to learn once you dive into the project 😄

s-gbz replied:
Since I don't have unix, I couldn't test myself if the variable is replaced, but it should work according to [this answer](https://askubuntu.com/a/76842). 

`# Set the deployment URL as parameter in the opensearch description`
`sed -i "s/placeholderURL/$deployedUrl/g" /path/to/opensearch.xml`

s-gbz replied:
Thanks for the suggestions, definetly going to integrate it somehow, really miss Linux 😅

gkalpak commented on aio/src/index.html:
```diff
@@ -4,7 +4,8 @@
   <title>Angular Docs</title>
   <base href="/">
-
+  <link rel="search" type="application/opensearchdescription+xml" href="assets/opensearch.xml" />
```
Nit: Why not put it next to all the `link` elements below? 😇  :grin:

s-gbz replied:
Fair point 👍

IgorMinar commented:
@s-gbz can you please rebase this PR on top of master? we changed some CI stuff and we won't get a ngbuilds.io preview on this PR until your PR is rebased. thanks!

mary-poppins commented:
You can preview ceedfc7 at https://pr25479-ceedfc7.ngbuilds.io/.

gkalpak commented on aio/angular.json:
```diff
@@ -61,7 +61,7 @@
                   "src": "src/environments/environment.ts",
                   "replaceWith": "src/environments/environment.next.ts"
-                }
+                }               
```
Oops 😱

gkalpak commented on aio/scripts/deploy-to-firebase.sh:
```diff
@@ -92,6 +92,10 @@ fi
 
+  # Set deployedUrl as parameter in the opensearch description
+  # deployedUrl must end with /
+  sed -i "s|placeholderURL|$deployedUrl|g" src/assets/opensearch.xml 
```
We should not change the `src/` file. Change the file in `dist/` instead.

gkalpak commented on aio/src/assets/opensearch.xml:
```diff
@@ -0,0 +1,9 @@
+<ShortName>Angular</ShortName>
+<Description>Search Angular</Description>
+<InputEncoding>UTF-8</InputEncoding>
+<Image width="16" height="16" type="image/x-icon">../assets/images/favicons/favicon-16x16.png</Image>
```
Why `../assets/images/...` and not just `images/...`?

gkalpak commented on aio/src/assets/opensearch.xml:
```diff
@@ -0,0 +1,9 @@
+<Description>Search Angular</Description>
+<InputEncoding>UTF-8</InputEncoding>
+<Image width="16" height="16" type="image/x-icon">../assets/images/favicons/favicon-16x16.png</Image>
+<Url type="text/html" method="get" template="placeholderURL?search={searchTerms}"/>
```
I would use something that stands out more as a placeholder (e.g. `<PLACEHOLDER_URL>` or something).

gkalpak commented on aio/src/index.html:
```diff
@@ -4,9 +4,10 @@
   <title>Angular Docs</title>
   <base href="/">
-
+  
```
😱 😱 😱

gkalpak commented on aio/src/index.html:
```diff
@@ -4,9 +4,10 @@
+  
   <meta name="viewport" content="width=device-width, initial-scale=1">
 
+  <link rel="search" type="application/opensearchdescription+xml" href="assets/opensearch.xml" />
```
Nit: I personally prefer ` />`, but the rest of the `link`s have `>` 😒

alfaproject replied:
`/>` is XHTML and I guess we are back to HTML (;

gkalpak replied:
Both forms are valid in HTML5. It is just a personal preference.
(I strongly prefer `/>`; most people don't 🤷‍♂️ But...consistency über alles 😛)

gkalpak reviewed (changes requested):
Looks reasonable :smiley:
(I left some minor comments.)

gkalpak commented on aio/src/assets/opensearch.xml:
```diff
@@ -0,0 +1,9 @@
+<Image width="16" height="16" type="image/x-icon">../assets/images/favicons/favicon-16x16.png</Image>
+<Url type="text/html" method="get" template="placeholderURL?search={searchTerms}"/>
+<moz:SearchForm>placeholderURL?search={searchTerms}</moz:SearchForm>
+</OpenSearchDescription>?
```
Bogus trailing `?`.

s-gbz replied:
Sorry looks like it's been a really tiring day last time 😄 

Hope it's alright now. Will fix asap if there's something you mind. (Just hope you don't 😂 )

s-gbz replied:
Also I remember about the --fixup command, will use it next time :)

mary-poppins commented:
You can preview 68a5b98 at https://pr25479-68a5b98.ngbuilds.io/.

IgorMinar reviewed (commented):
This looks much simpler than the original implementation! Good job!

Can you please rebase this PR against Master and let's get this merged.

s-gbz committed:
feat(docs-infra): Add opensearch description

Enables Chrome users to search angular.io and its subdomains from the browsers navigation bar.
Not sure if compatible with Firefox yet.
The queried term in the URL is removed after closing the search-results.

s-gbz commented on aio/scripts/deploy-to-firebase.sh:
```diff
@@ -92,6 +92,10 @@ fi
 
+  # Set deployedUrl as parameter in the opensearch description
+  # deployedUrl must end with /
+  sed -i "s|PLACEHOLDER_URL|$deployedUrl|g" dist/assets/opensearch.xml 
```
God damn, it worked on the Linux VM :@

Here goes a test if it works now.

gkalpak replied:
The reason it "doesn't work" is that the replacement happens in `deploy-to-firebase.sh` (i.e. only when deploying to production). We also need to do the replacement when deploying a PR preview; namely in [build-artifacts.sh][1]. But you have to build the correct domain, which is build off of the PR number and SHA:

```
https://pr<PR_NUMBER>-<SHORT_SHA>.ngbuilds.io/
```

, where `<PR_NUMBER>` can be taken from `$CIRCLE_PR_NUMBER` and `<SHORT_SHA>` from `${CIRCLE_SHA1:0:7}`.

[1]: https://github.com/angular/angular/blob/51c0d9cae9ac3972e819ae935d79145c6dae6aa7/aio/scripts/build-artifacts.sh

s-gbz replied:
Tested it on the machine. Let's see how it turns out.

mary-poppins commented:
You can preview c221690 at https://pr25479-c221690.ngbuilds.io/.

s-gbz commented:
@IgorMinar I squashed the commits, but saw your comment that it indeed didn't work (altough it did on the Linux VM). Here goes another try.. that doesn't work as well..

How do we proceed?

mary-poppins commented:
You can preview 07f7f94 at https://pr25479-07f7f94.ngbuilds.io/.

s-gbz committed:
fixup! feat(docs-infra): Add opensearch description

mary-poppins commented:
You can preview 29819ee at https://pr25479-29819ee.ngbuilds.io/.

gkalpak commented on aio/scripts/build-artifacts.sh:
```diff
@@ -6,11 +6,18 @@ source "`dirname $0`/../../scripts/ci/env.sh" print
 
 readonly INPUT_DIR=dist/
 readonly OUTPUT_FILE=$PROJECT_ROOT/$1
+readonly deployedUrl=https://pr${CIRCLE_PR_NUMBER}-${CIRCLE_SHA1:0:7}.ngbuilds.io/
```
Generally, it is a good idea to keep scripts CI-provider-independent (because updating them is a pain when switching or adding CI-providers).

Therefore, it would be better to pass `$CIRCLE_PR_NUMBER` and `$CIRCLE_SHA` as arguments in [.circleci/config.yml][1]. (That config file is already CircleCI-specific, so it's fine to use those vars there.)

[1]: https://github.com/angular/angular/blob/29819ee940ab7a849d2758d23b183a177caeeb78/.circleci/config.yml#L161

s-gbz replied:
Thanks for your suggestions bellow. Unfortunately I don't quite get what you'd like me to do here though 😅

gkalpak replied:
Sorry for being vague and thx for bearing with me 😇
So, here is what I suggested:

1. Change [this line][1] to pass the PR number and SHA to the script:

    ```diff
    -- run: ./aio/scripts/build-artifacts.sh $AIO_SNAPSHOT_ARTIFACT_PATH
    +- run: ./aio/scripts/build-artifacts.sh $AIO_SNAPSHOT_ARTIFACT_PATH $CIRCLE_PR_NUMBER $CIRCLE_SHA
    ```

2. In the `build-artifacts.sh` script, grab the PR number and SHA from the arguments. I.e. change [that line][2] to generate `$deployedUrl` like this:

    ```diff
    -readonly deployedUrl=https://pr${CIRCLE_PR_NUMBER}-${CIRCLE_SHA1:0:7}.ngbuilds.io/
    +readonly PR_NUMBER=$2
    +readonly PR_LAST_SHA=$3
    +readonly deployedUrl=https://pr${PR_NUMBER}-${PR_LAST_SHA:0:7}.ngbuilds.io/
    ```

This way, we can easily run `build-artifacts.sh` outside of CircleCI (e.g. locally or from another CI-provider), just by passing the PR number and SHA as arguments. I hope that makes sense :grin:

[1]: https://github.com/angular/angular/blob/29819ee940ab7a849d2758d23b183a177caeeb78/.circleci/config.yml#L161
[2]: https://github.com/angular/angular/blob/29819ee940ab7a849d2758d23b183a177caeeb78/aio/scripts/build-artifacts.sh#L9

gkalpak commented on aio/scripts/build-artifacts.sh:
```diff
@@ -6,11 +6,18 @@ source "`dirname $0`/../../scripts/ci/env.sh" print
+  
+  # Set deployedUrl as parameter in the opensearch description
+  # deployedUrl must end with /
+  sed -i "s|PLACEHOLDERURL|$deployedUrl|g" dist/assets/opensearch.xml
```
Since you are using this is multiple places, you could as well move it to a yarn script (and then you can even use `shelljs` and make this more Windows-friendly \o/).

It could like:

```json
"scripts": {
  ...
  "set-opensearch-url": "node --eval \"const sh = require('shelljs'); sh.set('-e'); sh.sed('-i', /PLACEHOLDERURL/g, process.argv[1], 'dist/assets/opensearch.xml');\"",
  ...
}
```

Then you can call it with `yarn set-opensearch-url $deployedUrl`.

gkalpak commented on aio/src/assets/opensearch.xml:
```diff
@@ -0,0 +1,9 @@
+<Description>Search Angular</Description>
+<InputEncoding>UTF-8</InputEncoding>
+<Image width="16" height="16" type="image/x-icon">images/favicons/favicon-16x16.png</Image>
+<Url type="text/html" method="get" template="PLACEHOLDERURL?search={searchTerms}"/>
```
Nit: I would rather use something more readable, like `PLACEHOLDER_URL` or something.

IgorMinar replied:
+1

gkalpak reviewed (changes requested):
Some final comments. But is definitely works now! Awesome job, @s-gbz 💯

s-gbz committed:
fixup! feat(docs-infra): Add opensearch description

mary-poppins commented:
You can preview cf1bc9c at https://pr25479-cf1bc9c.ngbuilds.io/.

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/package.json:
```diff
@@ -12,6 +12,7 @@
     "clean-dist": "node --eval \"require('shelljs').rm('-rf', 'dist')\"",
     "dev": "concurrently --kill-others --raw --success first \"yarn build-watch\" \"yarn test-watch\"",
     "lint": "tslint --project tsconfig.json",
+    "set-opensearch-url": "node --eval \"const sh = require('shelljs'); sh.set('-e'); sh.sed('-i', /PLACEHOLDER_URL/g, process.argv[1], 'dist/assets/opensearch.xml');\"",
```
That's the wrong `package.json`. [This is the one][1] you are looking for :grin:

<sub>In case you are wondering, this `package.json` (as everything inside `aio-builds-setup/`) is for the preview server, that we use to show angular.io previews for PRs ([docs][2]).</sub>

[1]: https://github.com/angular/angular/blob/cf1bc9c1ac6783c44d63d7a1d5fbf6875e44bc60/aio/package.json
[2]: https://github.com/angular/angular/blob/d6cd041cbd1a401554f14eed58a8a80e65c46570/aio/aio-builds-setup/docs/_TOC.md

s-gbz replied:
Are we going to have the command in both files, or shall I move it wholy to the other one?

gkalpak replied:
You are wrong 👉 
We only need the command in `aio/package.json` :grin:

gkalpak replied:
To add some more context: The preview server does not build the app; it downloads the build artifacts from CircleCI and just hosts them.

s-gbz replied:
Ok, then let's get this merged <3 Updated commit on the way

s-gbz committed:
fixup! feat(docs-infra): Add opensearch description

s-gbz committed:
fixup! feat(docs-infra): Add opensearch description

gkalpak reviewed (approved):
One final comment and this should be good to go :tada:

s-gbz committed:
fixup! feat(docs-infra): Add opensearch description

mary-poppins commented:
You can preview aea36d2 at https://pr25479-aea36d2.ngbuilds.io/.

gkalpak reviewed (approved):
Looks great! Thx, @s-gbz :clap:
Leaving it to @IgorMinar to give the final LGTM.

IgorMinar reviewed (approved):
looks good! thanks for all the updates

petebacondarwin commented on aio/src/app/app.component.ts:
```diff
@@ -384,6 +384,7 @@ export class AppComponent implements OnInit {
 
   hideSearchResults() {
     this.showSearchResults = false;
+    this.locationService.setSearch('', {});
```
Oh I see. You want to clear the site search from the URL if the user closes the results.
We need to be a bit more clever about this to prevent it clearing the other query params.
I'll send in a PR

petebacondarwin replied:
https://github.com/angular/angular/pull/26056
